### PR TITLE
Hash serialized scalars entirely; update test vectors to V10

### PR DESCRIPTION
**Pull request by conradoplg**
Based on #137 

Closes https://github.com/ZcashFoundation/frost/issues/138

### CODE REVIEW

**Comment by conradoplg** [comment]
@dconnolly: rebased and solved conflicts

File: frost-core/src/lib.rs [hunk 1]
```diff
@@ -173,28 +173,28 @@ pub trait Ciphersuite: Copy + Clone + PartialEq {
     ///
     /// Maps arbitrary inputs to non-zero `Self::Scalar` elements of the prime-order group scalar field.
     ///
-    /// [H1]: https://github.com/cfrg/draft-irtf-cfrg-frost/blob/master/draft-irtf-cfrg-frost.md#cryptographic-hash
+    /// [H1]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-10.html#name-cryptographic-hash-function
```
Comment (dconnolly) [inline frost-core/src/lib.rs:176]: Note to selves, we need to review these links each time we bump the version const

File: frost-core/src/frost/keys.rs [hunk 2]
```diff
@@ -430,30 +437,30 @@ pub struct PublicKeyPackage<C: Ciphersuite> {
...
 /// to be specified for this internal function for testability.
 ///
 /// Internally, [`generate_secret_shares`] performs verifiable secret sharing, which
 /// generates shares via Shamir Secret Sharing, and then generates public
 /// commitments to those shares.
 ///
 /// More specifically, [`generate_secret_shares`]:
-/// - Randomly samples of coefficients [a, b, c], this represents a secret
-/// polynomial f
+/// - Interpret [secret, coefficients[0], ...] as a secret polynomial f
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:452]:
```suggestion
/// - Interpret [secret, `coefficients[0]`, ...] as a secret polynomial f
```

File: frost-core/src/frost/keys.rs [hunk 3]
```diff
@@ -430,30 +437,30 @@ pub struct PublicKeyPackage<C: Ciphersuite> {
...
 /// generates shares via Shamir Secret Sharing, and then generates public
 /// commitments to those shares.
 ///
 /// More specifically, [`generate_secret_shares`]:
-/// - Randomly samples of coefficients [a, b, c], this represents a secret
-/// polynomial f
+/// - Interpret [secret, `coefficients[0]`, ...] as a secret polynomial f
 /// - For each participant i, their secret share is f(i)
-/// - The commitment to the secret polynomial f is [g^a, g^b, g^c]
+/// - The commitment to the secret polynomial f is [g^secret, g^coefficients[0], ...]
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:454]:
```suggestion
/// - The commitment to the secret polynomial f is [g^secret, `g^coefficients[0]`, ...]
```

File: frost-core/src/frost.rs [hunk 4]
```diff
@@ -244,7 +244,13 @@ where
                 let mut rho_input = vec![];
 
                 rho_input.extend_from_slice(&rho_input_prefix);
-                rho_input.extend_from_slice(&u16::from(c.identifier).to_be_bytes());
+                rho_input.extend_from_slice(
+                    <<C::Group as Group>::Field as Field>::serialize(
+                        // unwrap() is OK because this will become infallible after refactoring (#102)
+                        &c.identifier.to_scalar().unwrap(),
```
Comment (dconnolly) [inline frost-core/src/frost.rs:250]: This will look a little nicer when that other PR is merged

File: frost-core/src/frost.rs [hunk 5]
```diff
@@ -244,7 +244,13 @@ where
                 let mut rho_input = vec![];
 
                 rho_input.extend_from_slice(&rho_input_prefix);
-                rho_input.extend_from_slice(&u16::from(c.identifier).to_be_bytes());
+                rho_input.extend_from_slice(
+                    <<C::Group as Group>::Field as Field>::serialize(
+                        // unwrap() is OK because this will become infallible after refactoring (#102)
```
Comment (dconnolly) [inline frost-core/src/frost.rs:249]: 👍

File: frost-core/src/frost/keys.rs [hunk 6]
```diff
@@ -322,7 +323,13 @@ pub fn keygen_with_dealer<C: Ciphersuite, R: RngCore + CryptoRng>(
 
     let secret = SharedSecret::random(&mut rng);
     let group_public = VerifyingKey::from(&secret);
-    let secret_shares = generate_secret_shares(&secret, num_signers, threshold, rng)?;
+
+    let coefficients =
+        iter::repeat_with(|| <<C::Group as Group>::Field as Field>::random(&mut rng))
+            .take(threshold as usize - 1)
+            .collect();
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:330]: We've repeated this in tests, we may want to make it its own function

File: frost-core/src/frost/round1.rs [hunk 7]
```diff
@@ -259,22 +259,24 @@ pub struct GroupCommitmentShare<C: Ciphersuite>(pub(super) <C::Group as Group>::
...
     // TODO: AtLeastOne or other explicitly Sorted wrapper types?
     let mut sorted_signing_commitments = signing_commitments;
     sorted_signing_commitments.sort_by_key(|a| a.identifier);
 
     let mut bytes = vec![];
 
     for item in sorted_signing_commitments {
-        bytes.extend_from_slice(&u16::from(item.identifier).to_be_bytes()); // TODO: 2-bytes until spec moves off u16
+        bytes.extend_from_slice(
+            // unwrap() is OK because this will become infallible after refactoring (#102)
```
Comment (dconnolly) [inline frost-core/src/frost/round1.rs:276]: 👍

File: frost-core/src/frost/keys.rs (hunk 6 above)
Comment (conradoplg) [reply frost-core/src/frost/keys.rs]: Good idea, done in 4a7a9afd25eaeb9b215e58cc80605136a4608836

File: frost-core/src/frost/keys.rs [hunk 8]
```diff
@@ -322,7 +333,10 @@ pub fn keygen_with_dealer<C: Ciphersuite, R: RngCore + CryptoRng>(
 
     let secret = SharedSecret::random(&mut rng);
     let group_public = VerifyingKey::from(&secret);
-    let secret_shares = generate_secret_shares(&secret, num_signers, threshold, rng)?;
+
+    let coefficients = generate_coefficients::<C, R>(threshold as usize - 1, rng);
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:337]: 🎉

File: frost-core/src/frost/keys.rs [hunk 9]
```diff
@@ -13,6 +14,16 @@ use zeroize::{DefaultIsZeroes, Zeroize};
...
+/// Return a vector of randomly generated coefficients (Scalars).
+pub(crate) fn generate_coefficients<C: Ciphersuite, R: RngCore + CryptoRng>(
+    size: usize,
+    mut rng: R,
+) -> Vec<<<<C as Ciphersuite>::Group as Group>::Field as Field>::Scalar> {
+    iter::repeat_with(|| <<C::Group as Group>::Field as Field>::random(&mut rng))
+        .take(size)
+        .collect()
+}
+
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:26]: Fantastic 😎

File: frost-core/src/frost/keys.rs [hunk 10]
```diff
@@ -13,6 +14,16 @@ use zeroize::{DefaultIsZeroes, Zeroize};
 
 use crate::{frost::Identifier, Ciphersuite, Error, Field, Group, Scalar, VerifyingKey};
 
+/// Return a vector of randomly generated coefficients (Scalars).
```
Comment (dconnolly) [inline frost-core/src/frost/keys.rs:17]:
```suggestion
/// Return a vector of randomly generated polynomial coefficients ([`Scalar`]s).
```

**Review by dconnolly — APPROVED** [review]
💯

**Pull request merged**